Analysis/Utkscan/core/source/RootHandler.cpp

```diff
@@ -0,0 +1,111 @@
+/** \file RootHandler.cpp
+ * \brief Implemenation of class to dump event info to a root tree
+ * \author David Miller and S. V. Paulauskas
+ * \date Jan 2010
+ */
+#include <algorithm>
+#include <iostream>
+
+#include <TTree.h>
+
+#include "DetectorDriver.hpp"
+#include "RootHandler.hpp"
+
+using namespace std;
+
+RootHandler *RootHandler::instance_ = nullptr;
+TFile *RootHandler::file_ = nullptr; //!< The ROOT file that all the information will be stored in.
+std::set<TTree *> RootHandler::treeList_; //!< The list of trees known to the system
+std::map<unsigned int, TH1 *> RootHandler::histogramList_; //!< The list of 1D histograms known to the system
+
+/** Instance is created upon first call */
+RootHandler *RootHandler::get() {
+    if (!instance_)
+        instance_ = new RootHandler("histograms.root");
+    return (instance_);
+}
+
+RootHandler *RootHandler::get(const std::string &fileName) {
+    if (!instance_)
+        instance_ = new RootHandler(fileName);
+    return (instance_);
+}
+
+RootHandler::RootHandler(const std::string &fileName) {
+    file_ = new TFile(fileName.c_str(), "recreate");
+}
+
+RootHandler::~RootHandler() {
+    for (const auto &tree : treeList_) {
+        cout << "RootHandler::~RootHandler - Saving " << tree->GetEntries() << " tree entries to "
+             << tree->GetTitle() << endl;
+        tree->AutoSave();
+    }
+
+    if(file_) {
+        file_->Write(0, TObject::kOverwrite);
```

Comment:
Probably want `kWriteDelete` instead of `kOverwrite`. Also on line 110. 
https://root-forum.cern.ch/t/read-a-tfile-while-writing-with-another-process/18608/6?u=ksmith

Reply:
I found [this](https://root-forum.cern.ch/t/tfile-write/1590) forum post that said you could use either. What's the real difference? I couldn't find documentation.

Reply:
If I recall properly `kWriteDelete` didn't quite do what I was hoping, but if there's a good reason for it then I'm all ears.

Reply:
From the link I posted:
>`TObject::kOverwrite` is an explicit to first remove the previous copy and then write the new copy (lightly in place of the hold one) into the file.

Where as `kOverwrite` will simply write into the same memory block. This causes issues when the user is actively looking at the file. I'd suggest giving that thread a read if you haven't already. It discusses reading while writing which will most likely be needed for live monogramming.

The short documentation on these options: https://root.cern.ch/doc/master/classTObject.html#aaf76e963cc69ef877bb6c5af257e600caa84d56f05b42d86864444e7532b6b51b

Reply:
Ah, I see now that I read a little bit more. I think you're right, we want Overwrite.

Reply:
@ksmith0 : Based on what I'm seeing in your link, there's really nothing that's preventing us from viewing the output histogram online at all. Is there?